src/openrct2/actions/LandBuyRightsAction.hpp

```diff
@@ -0,0 +1,184 @@
+DEFINE_GAME_ACTION(LandBuyRightsAction, GAME_COMMAND_BUY_LAND_RIGHTS, GameActionResult)
+{
+private:
+    MapRange _range;
+    uint8_t _setting = static_cast<uint8_t>(LandBuyRightSetting::Count);
+
+    constexpr static rct_string_id _ErrorTitles[] = { STR_CANT_BUY_LAND, STR_CANT_BUY_CONSTRUCTION_RIGHTS_HERE };
+
+public:
+    LandBuyRightsAction()
+    {
+    }
+    LandBuyRightsAction(MapRange range, LandBuyRightSetting setting)
+        : _range(range)
+        , _setting(static_cast<uint8_t>(setting))
+    {
+    }
+
+    LandBuyRightsAction(CoordsXY coord, LandBuyRightSetting setting)
+        : _range(coord.x, coord.y, coord.x, coord.y)
+        , _setting(static_cast<uint8_t>(setting))
+    {
+    }
+
+    uint16_t GetActionFlags() const override
+    {
+        return GameAction::GetActionFlags();
+    }
+
+    void Serialise(DataSerialiser & stream) override
+    {
+        GameAction::Serialise(stream);
+
+        stream << DS_TAG(_range) << DS_TAG(_setting);
+    }
+
+    GameActionResult::Ptr Query() const override
+    {
+        return QueryExecute(false);
+    }
+
+    GameActionResult::Ptr Execute() const override
+    {
+        return QueryExecute(true);
+    }
+
+private:
+    GameActionResult::Ptr QueryExecute(bool isExecuting) const
+    {
+        auto res = MakeResult();
+
+        MapRange normRange = _range.Normalise();
+        // Keep big coordinates within map boundaries
+        auto aX = std::max<decltype(normRange.GetLeft())>(32, normRange.GetLeft());
+        auto bX = std::min<decltype(normRange.GetRight())>(gMapSizeMaxXY, normRange.GetRight());
+        auto aY = std::max<decltype(normRange.GetTop())>(32, normRange.GetTop());
+        auto bY = std::min<decltype(normRange.GetBottom())>(gMapSizeMaxXY, normRange.GetBottom());
+
+        MapRange validRange = MapRange{ aX, aY, bX, bY };
+
+        CoordsXYZ centre{ (validRange.GetLeft() + validRange.GetRight()) / 2 + 16,
+                          (validRange.GetTop() + validRange.GetBottom()) / 2 + 16, 0 };
+        centre.z = tile_element_height(centre.x, centre.y);
+
+        res->Position = centre;
+        res->ExpenditureType = RCT_EXPENDITURE_TYPE_LAND_PURCHASE;
+
+        // Game command modified to accept selection size
+        for (auto y = validRange.GetTop(); y <= validRange.GetBottom(); y += 32)
+        {
+            for (auto x = validRange.GetLeft(); x <= validRange.GetRight(); x += 32)
+            {
+                auto result = map_buy_land_rights_for_tile({ x, y }, isExecuting);
+                if (result->Error == GA_ERROR::OK)
+                {
+                    res->Cost += result->Cost;
+                }
+            }
+        }
+        if (isExecuting)
+        {
+            map_count_remaining_land_rights();
+        }
+        return res;
+    }
+
+    GameActionResult::Ptr map_buy_land_rights_for_tile(const CoordsXY loc, bool isExecuting) const
+    {
+        if (_setting >= static_cast<uint8_t>(LandBuyRightSetting::Count))
```

Comment:
I think this whole block can be moved to Query

Reply:
It already is in the Query. Its also in the execute though. I don't think it would read better if it had an additional if and there isn't much of a performance issue here either.